File: gpt-4-turbo/stage2/serial_construct_default_clause.cpp
#include <stdio.h>
#include <stdlib.h>
#include <openacc.h>

#ifndef T1
// T1:serial construct default clause, V:2.7-3.3
int test1(){
    int err = 0;
    srand(42); // Fixed seed for reproducibility

    int test_var = rand(); // Initialize a variable outside the OpenACC region
    int original_value = test_var; // Store its original value for comparison

    // Apply a simple operation in an OpenACC serial construct without specifying the default clause
    #pragma acc serial
    {
        test_var += 1; // Modify the variable inside the OpenACC region
    }

    // Check if the modification inside the OpenACC region affected the variable outside
    // Given that the default behavior typically shares variables between host and device,
    // we check if the change is visible here.
    // If the variable was private to the region (incorrectly due to some compiler issue), this check will fail.
    if(test_var != original_value + 1){
        err += 1; // Error if the variable was not correctly shared (or any unexpected behavior)
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed = 0;
    const int NUM_TEST_CALLS = 10;

#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif

    return failcode;
}

File: gpt-4-turbo/stage2/serial_construct_default_clause.f90
!T1:serial construct default clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        INTEGER, DIMENSION(10) :: array
        
        ! Populate array with initial values
        array = (/1, 2, 3, 4, 5, 6, 7, 8, 9, 10/)
        
        ! Applying serial construct with default(shared) to verify that
        ! modifications in the serial region are reflected outside.
        !$acc serial default(shared)
        array(5) = 50
        !$acc end serial
        
        ! errors = 0 indicates success; if array(5) != 50, the default(shared) failed
        IF (array(5) /= 50) THEN
          errors = errors + 1
        END IF

        ! Check errors
        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        LOGICAL :: test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM